### DISCUSSION TITLE
Change video name in SLP project by python

### DISCUSSION THREAD

**Discussion by deemok7** [talmolab/sleap#1343] [Help!]
Hi
Is it possible to change the video name or path in an SLP project using Python code? For instance, in a situation where the GUI cannot find the video and prompts us to locate it.

**Comment by deemok7** [comment]
I try this:
   project = sleap.load_file(filenameVal)
    project.videos.clear()
    # video_filenames = [os.path.basename(video.filename) for video in project.videos]
    new_video = sleap.Video.from_filename(newVideoPath)
    project.add_video(new_video)
    project.save(filenameVal)
    
it has error:

> ValueError: Video(backend=MediaVideo(filename='C:/Users/......\\118.mp4', grayscale=False, bgr=True, dataset='', input_format='channels_last')) is not in list

**Comment by roomrys** [comment]
Hi @mmmahdiii,

I am a bit confused by your question... Are you loading your project in the GUI or in the API? 

If you are loading your project in the GUI, then you will need to find your videos using the GUI pop-up (or exit, make some path changes/fixes, and reload). 

If you are loading your project through the API, then you can specify a list of search paths in the `Labels.load_file` function to help find your videos
```python
from sleap import load_file

ds = "path/to/slp"
path_to_videos_or_directories = ["path/to/video_0", "path/to/video/directory/for/videos_1_through_3"]
labels = load_file(ds, search_paths=path_to_videos_or_directories)
```

Let me know if this is what you were looking for.

Thanks,
Liezl

**Comment by roomrys** [comment]
Hi @mmmahdiii,

I saw you posted in #1332:
>Sorry for the delayed response. I had to rename all of the video file names, so the video names in the SLP file were different from their corresponding video paths. Therefore, I had to change the video names in the SLP project.

If you would rather not find all your videos in the GUI, you can change the video names programmatically (assuming the renaming followed some sort of pattern).

### Load your labels
```python
import sleap

ds = "path/to/slp"
path_to_videos_or_directories = ["path/to/video_0", "path/to/video/directory/for/videos_1_through_3"]

labels = sleap.load_file(ds, search_paths=path_to_video_directories)
```

### Access the video filenames
```python
videos = labels.videos

for vid in videos:
    print(vid.backend.filename)
```

### Replace the video filenames
Now that you know how to access the filenames, you can use either the order of the filenames (these won't change) or your renaming pattern to rename the videos within the project. Since I do not know your pattern, I just opted to example the order-based renaming
```python
new_names_in_correct_order = ["rename_video_0", "rename_video_1"]

for vid, new_name in zip(videos, new_names_in_correct_order):
    vid.backend.filename = new_name
```

### Save the new `Labels`
```python
new_ds = "path/for/new/slp"
labels.save(new_ds)
```

Thanks,
Liezl

**Reply by deemok7** [reply]
Hi @roomrys 
Let me clarify this further: I need to change the video name and path in SLP files using Python code, not GUI. I do not want to use search_paths. Could you please review my code and the error message again?
Thanks for response

**Reply by roomrys** [reply]
Hi @mmmahdiii,

The above answer holds after review, search paths are optional, but if you fully renamed the video, then they may lead to incorrect replacement as sometimes the video extension is the only thing used to determine if the video is "found". Just take out the `search_paths` and continue onto loading the labels and renaming.

Thanks,
Liezl
